src/hash_generator/dlNormalizer.js

```diff
@@ -0,0 +1,398 @@
+    return (`https://id.gs1.org/01/${addCheckDigitAndZeroPad(rawGTIN, throwError)}`);
+  }
+
+  if (!uri.match('^https?:(\/\/((([^\/?#]*)@)?([^\/?#:]*)(:([^\/?#]*))?))?((([^?#]*)(\/(01|gtin|8006|itip|8010|cpid|414|gln|417|party|8017|gsrnp|8018|gsrn|255|gcn|00|sscc|253|gdti|401|ginc|402|gsin|8003|grai|8004|giai)\/)([0-9]{4}[^\/]+)?[\/]?(([^?\n]*))?(#([^]*))?)|(\/[A-Za-z_-]{10}$))')) {
+    return uri;
+  }
+
+  if (uri.includes('?')) {
+    uri = uri.substring(0, uri.indexOf('?'));
+  }
+
+  uri = uri.split('/gtin/').join('/01/')
+    .split('/itip/').join('/8006/')
+    .split('/cpid/')
+    .join('/8010/')
+    .split('/gln/')
+    .join('/414/')
+    .split('/party/')
+    .join('/417/')
+    .split('/gsrnp/')
+    .join('/8017/')
+    .split('/gsrn/')
+    .join('/8018/')
+    .split('/gcn/')
+    .join('/255/')
+    .split('/sscc/')
+    .join('/00/')
+    .split('/gdti/')
+    .join('/253/')
+    .split('/ginc/')
+    .join('/401/')
+    .split('/gsin/')
+    .join('/402/')
+    .split('/grai/')
+    .join('/8003/')
+    .split('/giai/')
+    .join('/8004/')
+    .split('/cpv/')
+    .join('/22/')
+    .split('/lot/')
+    .join('/10/')
+    .split('/ser/')
+    .join('/21/');
+
+  if (!uri.match('^https:\/\/id.gs1.org\/(01|8006|8010|414|417|8017|8018|255|00|253|401|402|8003|8004)\/([0-9]{4}[^\/]+)(\/[^\/]+\/[^\/]+)?[\/]?(([^?\n]*))?(#([^\n]*))?|(\/[A-Za-z_-]{10}$)')) {
+    if (uri.indexOf('/00/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/00/'))}`;
+    else if (uri.indexOf('/01/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/01/'))}`;
+    else if (uri.indexOf('/253/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/253/'))}`;
+    else if (uri.indexOf('/255/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/255/'))}`;
+    else if (uri.indexOf('/401/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/401/'))}`;
+    else if (uri.indexOf('/402/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/402/'))}`;
+    else if (uri.indexOf('/414/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/414/'))}`;
+    else if (uri.indexOf('/417/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/417/'))}`;
+    else if (uri.indexOf('/8003/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8003/'))}`;
+    else if (uri.indexOf('/8004/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8004/'))}`;
+    else if (uri.indexOf('/8006/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8006/'))}`;
+    else if (uri.indexOf('/8010/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8010/'))}`;
+    else if (uri.indexOf('/8017/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8017/'))}`;
+    else if (uri.indexOf('/8018/') !== -1) uri = `https://id.gs1.org${uri.substring(uri.indexOf('/8018/'))}`;
+  }
+
+  if (!uri.match('^https:\/\/id.gs1.org\/01\/[0-9]{14}')) {
```

Comment:
Looks like this simply zero-pads a short GTIN/UPC. There should be a cleaner way than `else if`.

Reply:
I am not very familiar with regex. I'll look at this one once I have more time! :)